File: subscription-service/internal/domain/subscription.go
package domain

import (
	"time"

	"github.com/compendium-tech/compendium/subscription-service/internal/model"
	"github.com/google/uuid"
)

const (
	SubscriptionRolePayer  SubscriptionRole = "payer"
	SubscriptionRoleMember SubscriptionRole = "member"
)

type HandleUpdatedSubscriptionRequest struct {
	SubscriptionID string
	UserID         uuid.UUID
	Items          []SubscriptionItem
	Till           time.Time
	Since          time.Time
}

type SubscriptionItem struct {
	PriceID   string
	ProductID string
	Quantity  int
}

type SubscriptionRole string

type Subscription struct {
	Role    SubscriptionRole     `json:"role"`
	Tier    model.Tier           `json:"tier"`
	Since   time.Time            `json:"since"`
	Till    time.Time            `json:"till"`
	Members []SubscriptionMember `json:"members,omitempty"`
}

type SubscriptionResponse struct {
	IsActive      bool `json:"isActive"`
	*Subscription `json:"subscription,omitempty"`
}

type InvitationCodeResponse struct {
	InvitationCode *string `json:"invitationCode,omitempty"`
}

type SubscriptionMember struct {
	UserID          uuid.UUID        `json:"userId"`
	Name            string           `json:"name"`
	Email           string           `json:"email,omitempty"`
	Role            SubscriptionRole `json:"role,omitempty"`
	IsAccountActive bool             `json:"isAccountActive,omitempty"`
}

File: frontend-spa/src/api/subscription.ts
import { apiClient } from "./base"

export interface Subscription {
  role: SubscriptionRole
  tier: Tier
  since: string
  till: string
  members?: SubscriptionMember[]
}

export type SubscriptionRole = "payer" | "member"
export type Tier = "student" | "team" | "community"

export interface SubscriptionResponse {
  isActive: boolean
  subscription?: Subscription
}

export interface InvitationCodeResponse {
  invitationCode?: string
}

export interface SubscriptionMember {
  userId: string
  name: string
  email?: string
  role?: SubscriptionRole
  isAccountActive?: boolean
}

interface SubscriptionService {
  getSubscription: () => Promise<SubscriptionResponse>
  cancelSubscription: () => Promise<void>
  getSubscriptionInvitationCode: () => Promise<InvitationCodeResponse>
  updateSubscriptionInvitationCode: () => Promise<InvitationCodeResponse>
  removeSubscriptionInvitationCode: () => Promise<InvitationCodeResponse>
  joinSubscription: (invitationCode: string) => Promise<SubscriptionResponse>
  removeSubscriptionMember: (memberId: string) => Promise<void>
}

export const subscriptionService: SubscriptionService = {
  getSubscription: async () => {
    return (await apiClient.get("/subscription")).data
  },
  cancelSubscription: async () => {
    await apiClient.delete("/subscription")
  },
  getSubscriptionInvitationCode: async () => {
    return (await apiClient.get("/subscription/invitationCode")).data
  },
  updateSubscriptionInvitationCode: async () => {
    return (await apiClient.put("/subscription/invitationCode", {})).data
  },
  removeSubscriptionInvitationCode: async () => {
    return (await apiClient.delete("/subscription/invitationCode")).data
  },
  joinSubscription: async (invitationCode: string) => {
    return (
      await apiClient.post("/subscription/members/me", null, {
        params: { invitationCode },
      })
    ).data
  },
  removeSubscriptionMember: async (memberId: string) => {
    await apiClient.delete(`/subscription/members/${memberId}`)
  },
}
